micromanager_gui/_core_widgets/_property_widget.py

```diff
@@ -0,0 +1,344 @@
+# These widgets all implement PPropValueWidget for various PropertyTypes.
+# -----------------------------------------------------------------------
+
+T = TypeVar("T", bound=float)
+
+
+def _stretch_range_to_contain(wdg: QLabeledDoubleSlider, val: T) -> T:
+    """Set range of `wdg` to include `val`."""
+    if val > wdg.maximum():
+        wdg.setMaximum(val)
+    if val < wdg.minimum():
+        wdg.setMinimum(val)
+    return val
+
+
+class IntegerWidget(QSpinBox):
+    """Slider suited to managing integer values"""
+
+    def setValue(self, v: Any) -> None:
+        return super().setValue(_stretch_range_to_contain(self, int(v)))
+
+
+class FloatWidget(QDoubleSpinBox):
+    """Slider suited to managing float values"""
+
+    def setValue(self, v: Any) -> None:
+        # stretch decimals to fit value
+        dec = min(str(v).rstrip("0")[::-1].find("."), 8)
+        if dec > self.decimals():
+            self.setDecimals(dec)
+        return super().setValue(_stretch_range_to_contain(self, float(v)))
+
+
+class _RangedMixin:
+    _cast = int
+
+    # prefer horizontal orientation
+    def __init__(
+        self: QLabeledSlider,
+        orientation=Qt.Orientation.Horizontal,
+        parent: Optional[QWidget] = None,
+    ) -> None:
+        super().__init__(orientation, parent)
+
+    def setValue(self: QLabeledSlider, v: float) -> None:
+        return super().setValue(_stretch_range_to_contain(self, self._cast(v)))
+
+
+class RangedIntegerWidget(_RangedMixin, QLabeledSlider):
+    """Slider suited to managing ranged integer values"""
+
+
+class RangedFloatWidget(_RangedMixin, QLabeledDoubleSlider):
+    """Slider suited to managing ranged float values"""
+
+    _cast = float
+
+
+class IntBoolWidget(QCheckBox):
+    """Checkbox for boolean values, which are integers in pymmcore"""
+
+    valueChanged = Signal(int)
+
+    def __init__(self, parent: Optional[QWidget] = None) -> None:
+        super().__init__(parent)
+        self.toggled.connect(self._emit)
+
+    def _emit(self, state: bool):
+        self.valueChanged.emit(int(state))
+
+    def value(self) -> int:
+        return int(self.isChecked())
+
+    def setValue(self, val: Union[str, int]) -> None:
+        return self.setChecked(bool(int(val)))
+
+
+class ChoiceWidget(QComboBox):
+    """Combobox for props with a set of allowed values."""
+
+    valueChanged = Signal(str)
+
+    def __init__(self, allowed=(), parent: Optional[QWidget] = None) -> None:
+        super().__init__(parent)
+        self.currentTextChanged.connect(self.valueChanged.emit)
+        self._allowed = allowed
+        if allowed:
+            self.addItems(allowed)
+
+    def value(self) -> str:
+        return self.currentText()
+
+    def setValue(self, value: str) -> None:
+        value = str(value)
+        if value not in self._allowed:
+            raise ValueError(f"{value!r} must be one of {self._allowed}")
+        self.setCurrentText(value)
+
```

Comment:
@tlambert03 
can we add a method like this to the combobox?
```
def getValues(self) -> list:  # or def values(self)
        return[
            self.itemText(i) for i in range(self.count())
        ]
```

Reply:
we could, but what we really care about is the model (the cmmcore instance) not the view (the widget).   So, if you want a convenience method to get the allowed values from core (using `core.getAllowedPropertyValues` or `core.getStateLabels`), I'd propose that we make a convenience for accessing allowable values from core on the `PropertyWidget` class itself, rather than the getting the current values in the combobox.  sound ok?

Reply:
added `PropertyWidget.allowedValues`